### ISSUE TITLE
[entropy_src] Implementation & Documentation of Markov Test

### ISSUE THREAD

**Issue by martin-lueker** [lowRISC/opentitan#9759]
In the entropy_src the "Markov test" aims to detect either 
1. Oversampling of AST/RNG outputs leading to "clustered" input values that eventually change, but often are just repeats of the previous sample.  For example the string: "00111111000011000111000111000001111" has roughly equal numbers of 1's and 0's, but no good entropy source should generate such strings, because each bit is likely just a repeat of the previous one.
2. Wild oscillation of the RNG, in a distinctly non-random way.  For instance the string: "010101010101010101" has almost zero entropy, even though the number of 1's and 0's appears unbiased.

Detection accomplished by counting the number of changes in the a fixed number of RNG samples, and comparing number of "01"/"10" pairs to the number of "00"/"11" pairs.  Ideally the switched "01"vs. non-switched pairs should be roughly equal.

The [specification]() illustrates this test with a figure: 
![image](https://user-images.githubusercontent.com/47870387/146693810-1b46c562-1b78-4c24-a85c-3ece616393a5.png)

However the specification is very light on details and there are (at least) two ways of interpreting the spec:
1.  The total number of "10"/"01" pairs is counted and summed for all four RNG inputs.  In the figure above the test would output 5+2+5+2=14, and this value would be recorded for diagnostics and compared to thresholds.
2. The number of "10"/"01" pairs is counted for all for each RNG channel, and among each of the four channels the minimum and maximum value is taken, and compared to HI and LO threshold's.   The minimum and maximum would both be output separately for diagnostic purposes.

In verifying the design, I've noticed that the second approach has been adopted, even though the specification does not clearly state one way or the other.  This choice is suboptimal though, as the statistical power of this test is greatly reduced by only looking at the worst lines.   Individual lines can have more frequent outliers and the test thresholds need to be broadened to keep the number of false alarms within acceptable levels.

One could claim that this choice gives better insight into the behavior of a particular line, however such reasoning is not consistent with the construction of the other tests.  For instance, the adaptive proportion test sums statistics over all four lines.  This adaptive proportion test in principle could be implemented to detect defects in single RNG lines, taking min and max statistics across all four lines, but this is not what has been done in the implementation of the adaptive proportion test.

I recommend we:
1. Switch the implementation of the Markov Test to take the _sum_ over all four channels, to 
    a. get the most resolution from the statistics and 
    b. make the test more in line with the Adaptive Proportion test, and
2. Expand the specification to make our choices and motivations more clear (perhaps including some of the text above).

**Comment by martin-lueker** [comment]
For history, I can see now that this implementation choice has been in place since the [the very first commit of the Markov test](https://github.com/mwbranstad/opentitan/commit/cfd6d4181b92ab99d3a6024a76a8a2546aebfb74#diff-cd0b393e5e334df07777d36e8e8cd5965b924bb61245380c52356cd2d5f8c0da).

**Comment by zi-v** [comment]
@martin-lueker , I think the Markov choice is OK. 
However, I now see that the "adaptive proportion" ES test actually takes the first approach which I am not sure is ideal. Here is why:

From FIPS perspective, both adaptive proportion test (APT) and repetition count test (RCT) are mandatory (Markov is not).
With regards to FIPS cert to our RNG, one could try to claim that the RNG comprises four single bit sources or one 4-bit symbol source. This claim needs to be consistent with the way the APT and RCT are defined. I am not sure if we are allowed to choose our claim or if FIPS will mandate it for us, that's why we wanted to support both.

If we go with 4-bit symbol approach (which I believe is the mainstream), the APT maps to what we call the 'bucket' test, and the RCT is mapped to "repetition count symbol" test as described in the entropy source documentation.
If we eventually end up with the configuration of four sources, the RCT will be the "repetition count" but we'll have no matching ATP test, unless we change it to be channel by channel based.

To summarize, if there was something I would change right now in the entropy_src, it would be the adaptive proportion test.

**Comment by martin-lueker** [comment]
Hi @zi-v, 
  You bring up some really good points.   This is another place where we have conflicting goals in the requirements, be they due to differences in priorities, speculation about lab requirements, desire for optimizations etc.   We've managed the other challenges before, so I'm optimistic we can come up with something flexible here.  Getting a full analysis of the trade-offs will take effort and speculation about the properties of the RNG source, and if experience is any guide this is something we as a group could debate for a long time.
  What would you think about adding another MuBi field in the configuration register to select between these two behaviors?  Call it something like: "THRESHOLD_SCOPE".  This field would have to be selected at the same time as the thresholds, and it would modulate the behavior of _both_ Markov, and adaptive proportion tests.   Depending on the setting, the thresholds could be compared either on a per-channel basis or applied to the sum of all channels.
  The implement work involved would probably be light as we have examples of both.   It's one more thing to verify, but I'd be happy to do it.  
  If you ask nicely, I _might_ even be persuaded to make your preference the default. (!)  WDYT?

**Comment by zi-v** [comment]
@martin-lueker , I think it is good if you have the time and resources.
Otherwise, it is a risk management decision. Per my view, it is likely that we'll be able to take the 4-symbol approach, in which case, it will not matter much what we do with the four RNGs tests as long as it is reasonable. The problem is that I am not completely sure I am right :)

**Comment by martin-lueker** [comment]
Thanks @zi-v, "not being completely sure" seems par for the course for these decisions.   Which is all the more reason to put in software switching.

**Comment by tjaychen** [comment]
@martin-lueker do you think this is something we need a quick meeting to finalize?

**Comment by martin-lueker** [comment]
It wouldn't hurt.  @mwbranstad and I are prepared to put in the software swtich, however if anyone has any objections it would be good to hear them.

**Comment by tjaychen** [comment]
o if you guys are okay with the software switch already, let's go with that.
No need to have more meetings.

**Comment by tjaychen** [comment]
@martin-lueker 
@mwbranstad i think this is already addressed right..? assuming i remember correctly.

**Comment by martin-lueker** [comment]
No, not quite yet.
